core/src/main/java/feign/AsyncFeign.java

```diff
@@ -0,0 +1,328 @@
+ * <br><br>
+ * Target interface methods must return {@link CompletableFuture} with a non-wildcard type.
+ * As the completion is done by the {@link AsyncClient}, it is important that any subsequent
+ * processing on the thread be short - generally, this should involve notifying some other
+ * thread of the work to be done (for example, creating and submitting a task to an {@link ExecutorService}).
+ *
+ */
+public abstract class AsyncFeign<C> extends Feign {
+
+	public static <C> AsyncBuilder<C> asyncBuilder() {
+		return new AsyncBuilder<>();
+	}
+
+	private static class LazyInitializedExecutorService {
+
+		private static final ExecutorService instance = Executors.newCachedThreadPool(r -> {
+			Thread result = new Thread(r);
+			result.setDaemon(true);
+			return result;
+		});
+	}
+
+	public static class AsyncBuilder<C> {
+
+		private final Builder builder;
+		private Supplier<C> defaultContextSupplier = () -> null;
+		private AsyncClient<C> client;
+
+		private Logger.Level logLevel = Logger.Level.NONE;
+		private Logger logger = new NoOpLogger();
+
+		private Decoder decoder = new Decoder.Default();
+		private ErrorDecoder errorDecoder = new ErrorDecoder.Default();
+		private boolean decode404;
+		private boolean closeAfterDecode = true;
+
+		public AsyncBuilder() {
+			super();
+			this.builder = Feign.builder();								
+		}		
+		
+		public AsyncBuilder<C> defaultContextSupplier(Supplier<C> supplier) {
+			this.defaultContextSupplier = supplier;
+			return this;
+		}
+
+		public AsyncBuilder<C> client(AsyncClient<C> client) {
+			this.client = client;
+			return this;
+		}
+		
+		/**
+		 * @see Builder#mapAndDecode(ResponseMapper, Decoder)
+		 */
+		public AsyncBuilder<C> mapAndDecode(ResponseMapper mapper, Decoder decoder) {
+			this.decoder = (response, type) -> decoder.decode(mapper.map(response, type), type);
+			return this;
+		}
+		
+		/**
+		 * @see Builder#decoder(Decoder)
+		 */
+		public AsyncBuilder<C> decoder(Decoder decoder) {
+			this.decoder = decoder;
+			return this;
+		}
+
+		/**
+		 * @see Builder#decode404()
+		 */
+		public AsyncBuilder<C> decode404() {
+			this.decode404 = true;
+			return this;
+		}
+
+		/**
+		 * @see Builder#errorDecoder(ErrorDecoder)
+		 */
+		public AsyncBuilder<C> errorDecoder(ErrorDecoder errorDecoder) {
+			this.errorDecoder = errorDecoder;
+			return this;
+		}
+		
+		public AsyncBuilder<C> doNotCloseAfterDecode() {
+			this.closeAfterDecode = false;
+			return this;
+		}
+
+		public <T> T target(Class<T> apiType, String url) {
+			return target(new HardCodedTarget<>(apiType, url));
+		}
+		
+		public <T> T target(Class<T> apiType, String url, C context) {
+			return target(new HardCodedTarget<>(apiType, url), context);
+		}
+		
+		public <T> T target(Target<T> target) {
+			return build().newInstance(target);
+		}
+
+		public <T> T target(Target<T> target, C context) {
+			return build().newInstance(target, context);
+		}
+		
+		private static <C> AsyncClient<C> initDefaultClient() {
```

Comment:
Using a Builder, I don't see why we would need to have a static method to do this.  Methods like this are fine to be in the `build` method, given that it's being called locally.

Reply:
This could be inlined into the lazyInits() method right now, as long as the number of things going on there is small, it's probably clearer inlined.

Reply:
Inlined now